docs/adr/adr-014-bootstrap-from-previous-peers.md

```diff
@@ -0,0 +1,557 @@
+        opts ...Option[ClientParameters],
+ ) (*Exchange[H], error) {
+        params := DefaultClientParameters()
+        for _, opt := range opts {
+                opt(&params)
+        }
+ 
+        err := params.Validate()
+        if err != nil {
+                return nil, err
+        }
+ 
+        ex := &Exchange[H]{
+                host:       host,
+                protocolID: protocolID(params.networkID),
+                peerTracker: newPeerTracker(
+                        host,
+                        connGater,
++                       params.OnUpdatedPeers
++                       params.OnBlockedPeer,
+                ),
+                Params: params,
+        }
+ 
+        ex.trustedPeers = func() peer.IDSlice {
+                return shufflePeers(peers)
+        }
+        return ex, nil
+}
+ ```
+
+(_Code Snippet 2.b: Example of required changes to: go-header/p2p/exchange.go_)
+
+And then define `libhead.Exchange` options to set the event handlers on the `peerTracker`:
+
+```go
+func WithOnUpdatedPeers(f func([]]peer.ID)) Option[ClientParameters] {
+       return func(p *ClientParameters) {
+               switch t := any(p).(type) { //nolint:gocritic
+               case *PeerTrackerParameters:
+                       p.OnUpdatedPeers = f
+               }
+       }
+}
+
+func WithOnBlockedPeer(f func([]]peer.ID)) Option[ClientParameters] {
+       return func(p *ClientParameters) {
+               switch t := any(p).(type) { //nolint:gocritic
+               case *PeerTrackerParameters:
+                       p.OnBlockedPeer = f
+               }
+       }
+}
+```
+
+(_Code Snippet 2.c: Example of required changes to: go-header/p2p/go-header/p2p/options.go_)
+
+The event handlers to be supplied are callbacks that either:
+
+1. Put the new peer list into the peer store
+2. Remove a blocked peer from the peer store
+
+Such callbacks are easily passable as options at header module construction time, example:
+
+```go
+// newP2PExchange constructs a new Exchange for headers.
+func newPeerStore(cfg Config) func(ds datastore.Datastore) (PeerStore, error) {
+       return PeerStore(namespace.Wrap(ds, datastore.NewKey("peer_list")))
+}
+```
+
+(_Code Snippet 3.a: Example of required changes to: celestia-node/nodebuilder/p2p/misc.go_)
+
+```diff
++   fx.Provide(newPeerStore),
+    fx.Provide(newHeaderService)
+
+...
+
+- func(cfg Config, network modp2p.Network) []p2p.Option[p2p.ClientParameters] {
++ func(cfg Config, network modp2p.Network, peerStore datastore.Datastore) []p2p.Option[p2p.ClientParameters] {
+        return []p2p.Option[p2p.ClientParameters]{
+            ...
+            p2p.WithChainID(network.String()),
++           p2p.WithOnUpdatedPeers(func(peers []peer.ID) {
++               var topTen []peer.ID
++               if len(peers) >= 10 {
++                  topTen = peers[:9]
++               } else {
++                   topTen = peers
++               }
++               peerStore.Put(topTen)
```

Comment:
Here the top 10 are stored but above you say the sorting logic is optional - please add it above for consistency